aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,279 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	schema "github.com/m3db/m3aggregator/generated/proto/flush"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+type followerFlushManagerMetrics struct {
+	watchCreateErrors  tally.Counter
+	unmarshalErrors    tally.Counter
+	shardNotFound      tally.Counter
+	resolutionNotFound tally.Counter
+	kvUpdateFlush      tally.Counter
+	forcedFlush        tally.Counter
+}
+
+func newFollowerFlushManagerMetrics(scope tally.Scope) followerFlushManagerMetrics {
+	return followerFlushManagerMetrics{
+		watchCreateErrors:  scope.Counter("watch-create-errors"),
+		unmarshalErrors:    scope.Counter("unmarshal-errors"),
+		shardNotFound:      scope.Counter("shard-not-found"),
+		resolutionNotFound: scope.Counter("resolution-not-found"),
+		kvUpdateFlush:      scope.Counter("kv-update-flush"),
+		forcedFlush:        scope.Counter("forced-flush"),
+	}
+}
+
+type followerFlushManager struct {
+	sync.RWMutex
+
+	nowFn                 clock.NowFn
+	checkEvery            time.Duration
+	workers               xsync.WorkerPool
+	flushTimesKeyFmt      string
+	flushTimesStore       kv.Store
+	maxNoFlushDuration    time.Duration
+	forcedFlushWindowSize time.Duration
+	logger                xlog.Logger
+	scope                 tally.Scope
+
+	doneCh            <-chan struct{}
+	proto             schema.ShardSetFlushTimes
+	flushTimesKey     string
+	flushTimesUpdated bool
+	lastFlushed       time.Time
+	sleepFn           sleepFn
+	metrics           followerFlushManagerMetrics
+}
+
+func newFollowerFlushManager(
+	doneCh <-chan struct{},
+	opts FlushManagerOptions,
+) roleBasedFlushManager {
+	nowFn := opts.ClockOptions().NowFn()
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope()
+	return &followerFlushManager{
+		nowFn:                 nowFn,
+		checkEvery:            opts.CheckEvery(),
+		workers:               opts.WorkerPool(),
+		flushTimesKeyFmt:      opts.FlushTimesKeyFmt(),
+		flushTimesStore:       opts.FlushTimesStore(),
+		maxNoFlushDuration:    opts.MaxNoFlushDuration(),
+		forcedFlushWindowSize: opts.ForcedFlushWindowSize(),
+		logger:                instrumentOpts.Logger(),
+		scope:                 scope,
+		doneCh:                doneCh,
+		lastFlushed:           nowFn(),
+		sleepFn:               time.Sleep,
+		metrics:               newFollowerFlushManagerMetrics(scope),
+	}
+}
+
+func (mgr *followerFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	go mgr.watchFlushTimes()
+}
+
+// NB(xichen): no actions needed for initializing the follower flush manager.
+func (mgr *followerFlushManager) Init(*sync.RWMutex, []*flushBucket) {}
+
+func (mgr *followerFlushManager) Flush(
```

Comment:
The flushes are performed sequentially so they will not pile up. Specifically the follower flush manger will check for flush time updates from kv periodically (by default 1s) and only perform a flush if there is an update or sufficient time (by default 15 minutes) has passed without a flush. The leader by default persist flush times every 10s to reduce kv write volume, so if the follower takes more than 1s to finish the flush, that's totally fine because the next kv update from the leader won't come in for another 10s. Even if the follower takes more than 10s to flush, that's still okay because the follower will simply miss an intermediate kv flush times update and instead use the latest flush times for the next flush.